orc8r/gateway/go/status/gateway_status.go

```diff
@@ -147,43 +148,68 @@ func GetNetworkInfo() *NetworkInfo {
 		}
 		interfaces[i] = netIface
 	}
-	for i, rt := range hostRoutes {
-		dest := rt.Destination.IP.To4()
-		if dest == nil {
-			dest = rt.Destination.IP
-		}
-		gw := rt.Gateway.To4()
-		if gw == nil {
-			gw = rt.Gateway
-			if len(gw) == 0 {
-				if len(dest) == net.IPv4len {
-					gw = []byte{0, 0, 0, 0}
-				} else {
-					gw = net.IP([]byte{0, 0, 0, 0}).To16()
-				}
-			}
-		}
-		maskStr := rt.Destination.Mask.String()
-		if len(dest) == net.IPv4len {
-			maskV4 := net.IP(rt.Destination.Mask).To4()
-			if maskV4 != nil {
-				maskStr = maskV4.String()
-			}
+
+	return &NetworkInfo{
+		NetworkInterfaces: interfaces,
+		RoutingTable:      getRoutes(hostRoutes, true),
+	}
+}
+
+// getRoutes resolves the routing table based on the hostRoutes. This function
+// uses the source IP of the hostRoute to find the matching NetworkInterfaceId.
+// We use the LinkIndex to infer missing source IPs based on other host routes.
+// After resolving the unlinked host routes' source IPs, we make a recursive
+// call to getRoutes to add the missing entries to the routing table.
+func getRoutes(hostRoutes []netlink.Route, recursiveResolving bool) []*Route {
+	interfaceToIP := make(map[string]string)
+	var unlinkedHostRoutes []netlink.Route
+	var routes []*Route
+
+	for _, hostRoute := range hostRoutes {
+		src := getSourceIP(hostRoute)
+
+		if src == "" {
+			// No source IP is stored in hostRoute, we try to resolve this later
+			unlinkedHostRoutes = append(unlinkedHostRoutes, hostRoute)
+			continue
```

Comment:
Not sure if I'm missing something but if we build the map `linkIndexToIP` at the beginning of the function, couldn't we resolve the source IP here and then avoid the recursion?

```suggestion
			src = linkIndexToIP[hostRoute.LinkIndex]
```

Reply:
This isn't quite as easy, we need to fully populate the `interfaceToIP` map first to build `linkIndexToIP`. 

Since we do not know in which order we find the link indices, we always will need to iterate twice over the host routes. The first time we iterate to find all where the source IP is given, and the second time to try to find matching link indices for the host routes that have no stated source IP.

I chose recursion since it seemed like the simplest way to avoid redundant logic.

Reply:
I removed the recursion. It now looks also a bit different, so maybe have another look.